### PR TITLE
Generate additional fine-grained dependencies

### PR DESCRIPTION
**Pull request by JukkaL**
Now fine-grained dependencies can be generated for the majority of 
AST node types. Also add command line option `--dump-deps` that
prints all fine-grained dependencies to stdout.

Fine-grained incremental checking is still very much work in progress.
The additional AST node types aren't still supported end-to-end in 
fine-grained incremental checking; this only improves dependency 
generation.

### CODE REVIEW

File: mypy/server/deps.py [hunk 1]
```diff
@@ -101,14 +154,99 @@ def visit_import(self, o: Import) -> None:
...
+                object_type = self.type_map[lvalue.expr]
+                lvalue_type = self.type_map[lvalue]
+                type_triggers = get_type_dependencies(lvalue_type)
+                for attr_trigger in self.attribute_triggers(object_type, lvalue.name):
+                    for type_trigger in type_triggers:
+                        self.add_dependency(type_trigger, attr_trigger)
+        elif isinstance(lvalue, (ListExpr, TupleExpr)):
+            for item in lvalue.items:
+                self.process_lvalue(item)
+        # TODO: star lvalue
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Eventually, audit TODOs and turn them into asserts (but not yet).

File: mypy/server/deps.py [hunk 2]
```diff
@@ -101,14 +154,99 @@ def visit_import(self, o: Import) -> None:
...
+        self.process_lvalue(o.lvalue)
+        method = op_methods[o.op]
+        self.add_attribute_dependency_for_expr(o.lvalue, method)
+        if o.op in ops_with_inplace_method:
+            inplace_method = '__i' + method[2:]
+            self.add_attribute_dependency_for_expr(o.lvalue, inplace_method)
+
+    def visit_for_stmt(self, o: ForStmt) -> None:
+        super().visit_for_stmt(o)
+        self.add_attribute_dependency_for_expr(o.expr, '__iter__')
```
Comment (gvanrossum) [inline mypy/server/deps.py]: If there's no __iter__ depend on __getitem__.

File: mypy/server/deps.py [hunk 3]
```diff
@@ -101,14 +154,99 @@ def visit_import(self, o: Import) -> None:
...
+            self.add_type_dependencies(o.target_type)
+
+    def visit_print_stmt(self, o: PrintStmt) -> None:
+        super().visit_print_stmt(o)
+        if o.target:
+            self.add_attribute_dependency_for_expr(o.target, 'write')
+
+    def visit_del_stmt(self, o: DelStmt) -> None:
+        super().visit_del_stmt(o)
+        if isinstance(o.expr, IndexExpr):
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Could also be deleting an attribute (del x.y), should depend on that then.

File: mypy/server/deps.py [hunk 4]
```diff
@@ -101,14 +154,99 @@ def visit_import(self, o: Import) -> None:
...
         assert o.relative == 0  # Relative imports not supported
         for name, as_name in o.names:
-            assert as_name is None or as_name == name
             self.add_dependency(make_trigger(o.id + '.' + name))
 
+    def visit_block(self, o: Block) -> None:
+        if not o.is_unreachable:
+            super().visit_block(o)
+
     def visit_assignment_stmt(self, o: AssignmentStmt) -> None:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: PS. Traverser should also visit statement's .analyzed attribute.

File: mypy/server/deps.py [hunk 5]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+    def visit_type_application(self, e: TypeApplication) -> None:
+        super().visit_type_application(e)
+        for typ in e.types:
+            self.add_type_dependencies(typ)
+
+    def visit_index_expr(self, e: IndexExpr) -> None:
+        super().visit_index_expr(e)
+        self.add_indexing_method_dependency(e, lvalue=False)
+
+    def add_indexing_method_dependency(self, e: IndexExpr, lvalue: bool) -> None:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: The bool arg is borderline useful.

File: mypy/server/deps.py [hunk 6]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+        method = '__setitem__' if lvalue else '__getitem__'
+        self.add_operator_method_dependency(e.base, method)
+
+    def visit_unary_expr(self, e: UnaryExpr) -> None:
+        super().visit_unary_expr(e)
+        if e.op == '-':
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Maybe add a dict with the translation from operator to dunder method name?

File: mypy/server/deps.py [hunk 7]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+                self.add_operator_method_dependency(right, '__cmp__')
+
+    def process_binary_op(self, op: str, left: Expression, right: Expression) -> None:
+        method = op_methods.get(op)
+        if method:
+            if op != 'in':
+                self.add_operator_method_dependency(left, method)
+            else:
+                self.add_operator_method_dependency(right, method)
+            rev_method = reverse_op_methods.get(method)
```
Comment (gvanrossum) [inline mypy/server/deps.py]: This should only happen if op != 'in' -- maybe switch to op == 'in' logic.

File: mypy/server/deps.py [hunk 8]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+
+    def add_operator_method_dependency(self, e: Expression, method: str) -> None:
+        typ = self.type_map.get(e)
+        if typ is not None:
+            self.add_operator_method_dependency_for_type(typ, method)
+
+    def add_operator_method_dependency_for_type(self, typ: Type, method: str) -> None:
+        # Note that we operator methods can't be (non-metaclass) methods of type objects.
+        # TODO: TypedDict
+        # TODO: metaclasses
```
Comment (gvanrossum) [inline mypy/server/deps.py]: IOW TypeType or Callable representing a type.

File: mypy/server/deps.py [hunk 9]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+        self.add_iter_dependency(e.expr)
+
     # Helpers
 
     def add_dependency(self, trigger: str, target: Optional[str] = None) -> None:
+        """Add depedency from trigger to a target.
+
+        If the target is not given explicitly, use the current target.
+        """
         if target is None:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Move this after the trigger check.

File: mypy/server/deps.py [hunk 10]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
             target = self.current()
+        if trigger.startswith(('<builtins.', '<typing.')):
+            # Don't track dependencies to certain builtins to keep the size of
+            # the dependencies manageable. These dependencies should only
+            # change on mypy version updates, which will require a full rebuild
+            # anyway.
+            return
         self.map.setdefault(trigger, set()).add(target)
 
+    def add_type_dependencies(self, typ: Type, target: Optional[str] = None) -> None:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: TODO: Use this in more places where get_type_dependencies() + add_dependency() are called together.

File: mypy/server/deps.py [hunk 11]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+            return [make_trigger(member)]
+        elif isinstance(typ, FunctionLike) and typ.is_type_obj():
+            member = '%s.%s' % (typ.type_object().fullname(), name)
+            return [make_trigger(member)]
+        elif isinstance(typ, UnionType):
+            targets = []
+            for item in typ.items:
+                targets.extend(self.attribute_triggers(item, name))
+            return targets
+        else:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Verify we're not missing any cases. E.g. TypeType.

File: mypy/server/deps.py [hunk 12]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+        typ = self.type_map.get(e)
+        if typ is not None:
+            self.add_attribute_dependency(typ, name)
+
+    def add_iter_dependency(self, node: Expression) -> None:
+        typ = self.type_map.get(node)
+        if typ:
+            self.add_attribute_dependency(typ, '__iter__')
+
     def push(self, component: str) -> str:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Consider refactoring all the push/enter methods. Or push -> push_function?

File: mypy/server/deps.py [hunk 13]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+
+    def leave_scope(self) -> None:
+        """Leave the innermost scope."""
+        self.scope_stack.pop()
+
+    def current_scope(self) -> Optional[Node]:
+        return self.scope_stack[-1]
+
 
 def get_type_dependencies(typ: Type) -> List[str]:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: Add a docstring. And maybe rename to get_type_triggers?

File: mypy/server/deps.py [hunk 14]
```diff
@@ -229,9 +553,31 @@ def visit_uninhabited_type(self, typ: UninhabitedType) -> List[str]:
...
+
+
+def dump_all_dependencies(modules: Dict[str, MypyFile],
+                          type_map: Dict[Expression, Type],
+                          python_version: Tuple[int, int]) -> None:
+    """Generate dependencies for all interesting modules and print them to stdout."""
+    all_deps = {}  # type: Dict[str, Set[str]]
+    for id, node in modules.items():
+        # print('processing', id)
+        if id in ('builtins', 'typing') or 'typeshed' in node.path:
```
Comment (gvanrossum) [inline mypy/server/deps.py]: /typeshed/ ?

File: mypy/server/deps.py [hunk 15]
```diff
@@ -1,33 +1,47 @@
...
+                               type_map: Dict[Expression, Type],
+                               python_version: Tuple[int, int]) -> Dict[str, Set[str]]:
     """Get dependencies of a target -- don't recursive into nested targets."""
-    visitor = DependencyVisitor(prefix, type_map)
+    prefix = module_id
+    surrounding_target = module_id
+    scope = None
+    if isinstance(node, FuncBase) and node.info:
+        # TODO: Nested classes
+        # TOOD: Add tests
```
Comment (ilevkivskyi) [inline mypy/server/deps.py]: Typo TOOD -> TODO

File: mypy/server/deps.py [hunk 16]
```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
...
+            if rev_method:
+                self.add_operator_method_dependency(right, rev_method)
+
+    def add_operator_method_dependency(self, e: Expression, method: str) -> None:
+        typ = self.type_map.get(e)
+        if typ is not None:
+            self.add_operator_method_dependency_for_type(typ, method)
+
+    def add_operator_method_dependency_for_type(self, typ: Type, method: str) -> None:
+        # Note that we operator methods can't be (non-metaclass) methods of type objects.
```
Comment (ilevkivskyi) [inline mypy/server/deps.py]: Typo: we -> the(?)

File: mypy/server/deps.py (hunk 4 above)
Comment (JukkaL) [reply mypy/server/deps.py]: Turns out that `AssignmentStmt` doesn't have `.analyzed`.

File: mypy/server/deps.py (hunk 3 above)
Comment (JukkaL) [reply mypy/server/deps.py]: This was working already due to the `super()` traversal method.

**Pull request merged**